src/util/query_exec_util.cpp

```diff
@@ -0,0 +1,254 @@
+                                               util->optimizer_timeout_);
+}
+
+QueryExecUtil::QueryExecUtil(common::ManagedPointer<transaction::TransactionManager> txn_manager,
+                             common::ManagedPointer<catalog::Catalog> catalog,
+                             common::ManagedPointer<settings::SettingsManager> settings,
+                             common::ManagedPointer<optimizer::StatsStorage> stats, uint64_t optimizer_timeout)
+    : txn_manager_(txn_manager),
+      catalog_(catalog),
+      settings_(settings),
+      stats_(stats),
+      optimizer_timeout_(optimizer_timeout) {}
+
+void QueryExecUtil::ClearPlans() {
+  schemas_.clear();
+  exec_queries_.clear();
+}
+
+void QueryExecUtil::SetDatabase(catalog::db_oid_t db_oid) {
+  if (db_oid != catalog::INVALID_DATABASE_OID) {
+    db_oid_ = db_oid;
+  } else {
+    auto *txn = txn_manager_->BeginTransaction();
+    db_oid_ = catalog_->GetDatabaseOid(common::ManagedPointer(txn), catalog::DEFAULT_DATABASE);
+    txn_manager_->Commit(txn, transaction::TransactionUtil::EmptyCallback, nullptr);
+  }
+}
+
+void QueryExecUtil::UseTransaction(catalog::db_oid_t db_oid,
+                                   common::ManagedPointer<transaction::TransactionContext> txn) {
+  NOISEPAGE_ASSERT(!own_txn_, "QueryExecUtil already using a transaction");
+  SetDatabase(db_oid);
+  txn_ = txn.Get();
+  own_txn_ = false;
+}
+
+void QueryExecUtil::BeginTransaction(catalog::db_oid_t db_oid) {
+  NOISEPAGE_ASSERT(txn_ == nullptr, "Nesting transactions not supported");
+  SetDatabase(db_oid);
+  txn_ = txn_manager_->BeginTransaction();
+  own_txn_ = true;
+}
+
+void QueryExecUtil::EndTransaction(bool commit) {
+  NOISEPAGE_ASSERT(txn_ != nullptr, "Transaction has not started");
+  NOISEPAGE_ASSERT(own_txn_, "EndTransaction can only be called on an owned transaction");
+  if (commit)
+    txn_manager_->Commit(txn_, transaction::TransactionUtil::EmptyCallback, nullptr);
+  else
+    txn_manager_->Abort(txn_);
+  txn_ = nullptr;
+  own_txn_ = false;
+}
+
+std::pair<std::unique_ptr<network::Statement>, std::unique_ptr<planner::AbstractPlanNode>> QueryExecUtil::PlanStatement(
+    const std::string &query, common::ManagedPointer<std::vector<parser::ConstantValueExpression>> params,
+    common::ManagedPointer<std::vector<type::TypeId>> param_types, std::unique_ptr<optimizer::AbstractCostModel> cost) {
+  NOISEPAGE_ASSERT(txn_ != nullptr, "Transaction must have been started");
+  auto txn = common::ManagedPointer<transaction::TransactionContext>(txn_);
+  auto accessor = catalog_->GetAccessor(txn, db_oid_, DISABLED);
+
+  std::unique_ptr<network::Statement> statement;
+  try {
+    std::string query_tmp = query;
+    auto parse_tree = parser::PostgresParser::BuildParseTree(query_tmp);
+    statement = std::make_unique<network::Statement>(std::move(query_tmp), std::move(parse_tree));
+  } catch (std::exception &e) {
+    // Catched a parsing error
```

Comment:
I don't know whether we want to crash the system here, but at least we should print out those exceptions with `LOG_ERROR`. Otherwise, it's pretty difficult to debug (e.g., when I'm trying to run the pilot with the new task manager today).